docs/best-practices/retaining-correct-variable-types.md

```diff
@@ -1,6 +1,40 @@
 ---
 title: Retaining Correct Variable Types
 ---
+### Adding Type Annotations to Your Resolver
+`dara.core.interactivity.plain_variable.Variable`s are stored in the browser and `dara.core.interactivity.derived_variable.DerivedVariable`s and `py_component`s are calculated on the server. When a `Variable` is passed to one of these, it is serialized into a format that FastApi can handle and becomes a python basic type format on the server side. Eg, a  `numpy.array` will became a  `list`, a `numpy.int8` will become a `int`.
```

Comment:
Technically the serialization in `Variable` happens when the `Variable` is included anywhere in your template, not just in a DV or py_component, e.g.

`Text(text=Variable('some text'))` -> Dara will attempt to serialize the variable alongside the default.

Maybe worth simplifying this part a little bit. Here's what I'd suggest:


> `dara.core.interactivity.plain_variable.Variable`s are stored in the browser and `dara.core.interactivity.derived_variable.DerivedVariable`s and `py_component`s are calculated on the server. This means that there are times when Dara needs to be able to serialize and deserialize your values.
> 
> Static values or `Variable` default values need to be serialized to JSON as part of the component tree - either when included directly in the page or when returned from a `py_component`.
> 
> When a user-defined resolver (for `DerivedVariable`s, `py_component`s or action) is called, Dara receives the `Variable` values from the frontend as the serialized JSON. This means your data structures would be turned into primitive types, a `numpy.array` will became a `list`, a `numpy.int8` will become a `int`. 
> 
> To preserve your original types, Dara provides an encoder system which will attempt to turn the primitive types back into the original type, as long as you annotate your resolver with a type for which an encoder is defined.